File: src/algorithm/leetcode/Problem53.java
package algorithm.leetcode;

public class Problem53 {
    public static int maxSubArray(int[] nums) {
        int n = nums.length;
        if (n == 0)
            return 0;
        if (n == 1)
            return nums[0];
        int[] sum = new int[n];
        sum[0] = nums[0];
        int maxSum = nums[0];
        for (int i = 1; i < n; i++) {
            sum[i] = Math.max(sum[i - 1] + nums[i], nums[i]);
            maxSum = Math.max(maxSum, sum[i]);
        }
        return maxSum;
    }
}

File: src/algorithm_cpp/problem53.cpp
class Solution {
public:

    int maxSubArray(vector<int> & nums) {
	int dp[nums.size()];
	dp[0] = nums[0];
	for (int i = 0; i < nums.size(); i++) {
	    dp[i] = 0;
	}
	dp[0] = nums[0];
	int max = nums[0];
	for (int i = 1; i < nums.size(); i++) {
	    dp[i] = dp[i - 1] + nums[i] > nums[i] ? dp[i - 1] + nums[i] : nums[i];
	    if (dp[i] > max)
		max = dp[i];
	}
	return max;
    }
};
